src/backends/perf_pt/perf_pt.c

```diff
@@ -105,81 +105,66 @@ struct tracer_thread_args {
+    void                **trace_buf;        // Pointer to the buffer to copy the trace into.
+    __u64               trace_bufsize;      // Initial capacity of the trace buffer (in bytes).
+    __u64               *trace_len;         // Pointer to the trace length (in bytes).
 };
 
 
 // Private prototypes.
-static bool write_buf_to_disk(int, void *, __u64);
-static bool read_circular_buf(void *, __u64, __u64, __u64 *, int);
-static bool poll_loop(int, int, int, struct perf_event_mmap_page *, void *);
+static bool read_aux(void *, __u64, __u64, __u64 *, void **, __u64, __u64 *);
+static bool poll_loop(int, int, struct perf_event_mmap_page *, void *, void **,
+                      __u64, __u64 *);
 static void *tracer_thread(void *);
 static int open_perf(pid_t target_tid);
 
 // Exposed Prototypes.
 struct tracer_ctx *perf_pt_start_tracer(struct tracer_conf *);
-int perf_pt_stop_tracer(struct tracer_ctx *tr_ctx);
-
+int perf_pt_stop_tracer(struct tracer_ctx *tr_ctx, uint8_t **, __u64 *);
+void *perf_pt_trace_buf(struct tracer_ctx *tr_ctx, uint32_t *len);
 
 /*
- * Write part of a buffer to a file descriptor.
+ * Read data out of the AUX buffer.
  *
- * The implementation is a little convoluted due to friction between ssize_t
- * and __u64.
- *
- * Returns true on success or false otherwise.
- */
-static bool
-write_buf_to_disk(int fd, void *buf, __u64 size) {
-    char *buf_c = (char *) buf;
-    size_t block_size = SSIZE_MAX;
-
-    while (size > 0) {
-        if (block_size > size) {
-            block_size = size;
-        }
-        ssize_t res = write(fd, buf, block_size);
-        if (res == -1) {
-            if (errno == EINTR) {
-                // Write interrupted before anything written.
-                continue;
-            }
-            return false;
-        }
-        size -= res;
-        buf_c += res;
-    }
-    return true;
-}
-
-/*
- * Read data out of a circular buffer.
+ * Reads `size` bytes from `aux_buf` into `trace_buf`, updating `*trace_len`.
  *
  * Returns true on success and false otherwise.
  */
 static bool
-read_circular_buf(void *buf, __u64 size, __u64 head_monotonic, __u64 *tail_p, int out_fd) {
+read_aux(void *aux_buf, __u64 size, __u64 head_monotonic, __u64 *tail_p,
+         void **trace_buf, __u64 trace_bufsize, __u64 *trace_len) {
     __u64 head = head_monotonic % size; // Head must be manually wrapped.
     __u64 tail = *tail_p;
 
+    // First check we won't overflow the (destination) trace buffer.
+    __u64 new_data_size;
     if (tail <= head) {
         // No wrap-around.
-        DEBUG("read with no wrap");
-        if (!write_buf_to_disk(out_fd, buf + tail, head - tail)) {
-            return false;
-        }
+        new_data_size = head - tail;
     } else {
         // Wrap-around.
+        new_data_size = (size - tail) + head;
+    }
+
+    // For now we simply crash out if we exhaust the buffer.
+    DEBUG("current trace length: %lld", *trace_len);
```

Comment:
Is this debugging useful still? If not, I'm keen on having less output when possible.

Reply:
It's useful for development yes, but we can define `DEBUG` empty for normal usage.

Reply:
What I'm saying is that there's always a temptation to leave lots of print statements (which this is) lurking around that are no longer useful. I'm very keen not to leave anything around that isn't still actively useful. If we're not careful we'll end up like one of those programs that spews stuff to the terminal forever, because no-one's brave enough to remove the print statements.

Reply:
If we define `DEBUG` empty, nothing is printed. I would like to keep the DEBUG statements in.

Reply:
We have to be really careful. 95% of debugging statements left in programs are not only not useful, but they actively make finding real problems harder. Here's one guide I sometimes use: would a normal user ever want to see this output? Could it be useful to them? If not, we should remove them. Remember, adding print statements in for temporary debugging is super cheap.

Reply:
> would a normal user ever want to see this output? 

No, but they wouldn't anyway. It's like `DNPRINTF` in the OpenBSD kernel. You turn it on at compile time if you want it.

If you really feel very strongly, we can kill it, but I'm telling you, I do find this code very useful.

Reply:
At the very least, let's think about reducing the output. For example, the two lines that come after this one, seem superfluous because the `assert` catches any violations, and they seem to be the result of defensive programming. I'm not sure the "read with wrap/no wrap" is useful either, but maybe I'm wrong.

Reply:
That sounds more reasonable.

I think the main goal of these `DEBUG`s is to know roughly where to start looking without having to do a binary chop to find out.

Perhaps I could try to limit the statements to one at the top of important-ish functions? Would you be OK with that?

Reply:
I wouldn't like to put that sort of rule in place. I'm really not keen on these statements lurking around before we know they might be useful. They clog up code something rotten and, when done pre-emptively, tend not to produce very useful output IME. More is not always better ;)

Reply:
I've killed all of the calls to `DEBUG`, but left the macro in, since I don't want to have to write that afresh every time.